asapdiscovery-alchemy/asapdiscovery/alchemy/predict.py

```diff
@@ -989,3 +989,76 @@ def download_cdd_data(protocol_name: str) -> pd.DataFrame:
+
+
+def clean_result_network(network, console=None):
+    """
+    Cleans an incoming result network JSON file from some issues that might occur. Current procedures:
+    - removes edges that have DG==0.0, this happens when there is inconsistent stereo annotation in input ligands
+    such that after stereo enumeration there are duplicate ligands.
+    - cleans imbalanced complex/solvent legs, e.g. when some have failed or when stereo expansion was done unintionally
+
+    returns the loaded FreeEnergyCalculationNetwork.
+    """
+    import json
+    import sys
+    import tempfile
+
+    import numpy as np
+    import rich
+    from asapdiscovery.alchemy.schema.fec import FreeEnergyCalculationNetwork
+    from rich.padding import Padding
+
+    with open(network) as f:
+        d = json.load(f)
+        input_results = d["results"]["results"]
+
+    # 1. remove edges where DG is 0.0
+    cleaned_results = []
+    for result in input_results:
+        if not result["estimate"]["magnitude"] == 0.0:
+            cleaned_results.append(result)
+
+    num_0_0_removed = len(input_results) - len(cleaned_results)
+
+    # 2. balance between complex/solvent replicates, such that n=N=1
+    deduped_results_dict = (
+        {}
+    )  # make a dict first that we can query; if an edge has multiple results it'll be a list of results
+    for result in cleaned_results:
+        transform = f"{result['ligand_a']}~{result['ligand_b']}_{result['phase']}"
+        if transform in deduped_results_dict:
+            deduped_results_dict[transform].append(result)
+        else:
+            deduped_results_dict[transform] = [result]
+
+    deduped_results = []
+    for _, results in deduped_results_dict.items():
+        if len(results) > 1:
+            # take the arithmetic mean of DG and dDG and add the replaced first result,
+            # all provenance data is constant between these repeats anyway
+            mean_DG = np.mean([result["estimate"]["magnitude"] for result in results])
+            mean_dDG = np.mean(
+                [result["uncertainty"]["magnitude"] for result in results]
+            )
+            results[0]["estimate"]["magnitude"] = mean_DG
+            results[0]["uncertainty"]["magnitude"] = mean_dDG
+
+        deduped_results.append(results[0])
+
+    num_dupes_removed = len(cleaned_results) - len(deduped_results)
+    if console:
+        message = Padding(
+            f"Cleaned incoming result network. Removed {num_0_0_removed} edges with DG==0.0 kcal/mol and removed {num_dupes_removed} edges to balance between complex/solvent replicates.",
+            (1, 0, 1, 0),
+        )
+        console.print(message)
+
+    # replace the original results in the JSON dict with the new cleaned results and load in
+    # the FECNetwork using a temp file
+    d["results"]["results"] = deduped_results
+    with tempfile.NamedTemporaryFile(suffix=".json") as fp:
+        with open(fp.name, "w") as f:
+            json.dump(d, f)
```

Comment:
Dont use tempfile here, just use `FreeEnergyCalculationNetwork.parse_obj(dict)`

Reply:
ooh I didn't know that was an option. That gives a ton of validation errors though:
```
  File "/Users/jenkescheen/asapdiscovery/asapdiscovery-alchemy/asapdiscovery/alchemy/predict.py", line 1060, in clean_result_network
    return FreeEnergyCalculationNetwork.parse_obj(d)
  File "pydantic/main.py", line 526, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1019 validation errors for FreeEnergyCalculationNetwork
```
will open an issue after merging to add this later on